### PR TITLE
Database seeding

### LINKED ISSUE #478: Seed the database - For testing purposes

### PR DESCRIPTION
**Pull request by ErlendHaa**
Closes #478

### CODE REVIEW

**Review by achaikou — COMMENTED** [review]
Looks great!
Do not have enough knowledge to thoroughly verify what the seed file does, but if something we can fix any occurring issues when writing actual tests.

File: frontend/cypress/support/mocks.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+
+interface IAction {
+    questionOrder: number
+    assignedTo: Participant
+    createdBy: Participant
+    dueDate: Date
+    title: string
+    priority: Priority
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply frontend/cypress/support/mocks.ts:71]: 0. Future prs, yes 👍 
1. Wouldn't put it here, as that would mean random by-default, and you would have to explicitly give all arguments to opt-out.  We could make a set of functions that makes random Answer, Action etc _in addition_ to what we have. But one thing to be vary about is the possibility of the randomness to create unrealistic scenarios
2. Not sure it's better or worse, don't see any obvious benefit of either. So gonna stick with default values. But thanks for pointing out builder pattern as a possible solution

File: frontend/cypress/support/gql.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        progressEvaluation(
+            evaluationId: $evaluationId,
+            newProgression: $newProgression
+        ) {
+            id
+        }
+    }
+`
+
+export const PROGRESS_PARTICIPANT = `
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply frontend/cypress/support/gql.ts:59]: Good that you pointed this out, because I got it a bit wrong. Not in the code, but I though of it wrong. So in the UI we have 2 elements that does progression. The complete switch and the Finish button. The complete switch is individual (pr user) and is hooked up to ProgressParticipant. The Finish button is only for the facilitator and progresses the entire evaluation _and_ all participants. I.e. is syncronizes everyone to the same level.
Bottom line:
> A participant can be either on the same progression, or one progression level ahead of the evaluation. And it can never lag behind. 

So the implementation is still correct, but it can be optimised a little

File: frontend/cypress/support/evaluationSeed.ts [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+                    }
+
+                }
+            )
+        }
+        ).then( (res) => {
+            action.id = res.body.data.createAction.id
+        })
+    }
+    ).then( () => { return seed.notes }
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply frontend/cypress/support/evaluationSeed.ts:338]: It's not this that caused the weird signature, it's the early return when there is no summary. 

And to the `return seed.notes` etc:

It's the combination of `.then` and `.each` that is interesting. When `.then` returns a list of notes, `.each` is able to capture each note separately and execute the query for that note. So it's behaves like a for-loop, but with some extra benefits wrt to concurrency:
1: all the note queries are executed async
2: `.each` will not return (progress to the next `.then`) until _all_ note queries are finished.  
3: `.each` stitched together all responses and returns a list of note-responses to the _next_ `.then`

Why is this important?
In the Action/Note case we know that a note is attached to an action, hence we must be sure that the action is created before we start posting notes. then + each is perfect for this as we can post all actions async, then wait till all of them are done before moving to notes

File: frontend/cypress/support/evaluationSeed.ts [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+ * -------
+ *
+ * The responsibility for creating sensible Evaluation seeds are left to the
+ * user of this class.  This class _does not_ sanitize the evaluation state
+ * that's being set up. This means it's perfectly possible to create illogical
+ * states. E.g. you can crate an Action even tho the Evaluation - and all the
+ * participants,  are at progression.Individual. The rationale for this is that
+ * implementing sanitizers will basically mean re-implementing the behaviour of
+ * the application, which is error-prone and can easily get out-of-sync.
+ */
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply frontend/cypress/support/evaluationSeed.ts:99]: > I am missing in the docs one important explicit thing: is the database forcefully cleared or not?
`... And then each test that is using the preset can simply use call .plant() to create its very own Evaluation in the database.`

On CI:
Definitely a in-mem db. And I see no need to clear it (except the one case you mentioned)
Let's discuss over a coffee when you wake up :D

File: frontend/cypress/support/evaluationSeed.ts [hunk 5]
```diff
@@ -0,0 +1,375 @@
+import { Progression, Severity, Priority, Question, Role } from '../../src/api/models'
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply frontend/cypress/support/evaluationSeed.ts:1]: I followed /src. Seems we are already at chaos

File: frontend/cypress/support/auth.ts [hunk 6]
```diff
@@ -102,7 +112,7 @@ declare global {
              * Retrieve user token from mock-oauth2-server and store it
              * @example cy.login(extHire1)
              */
-            login(value?: User): void
+            login(value?: User): Cypress.Chainable
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply frontend/cypress/support/auth.ts:115]: We will do that in the potential future pr where we cache user tokens ourselves

**Review by achaikou — APPROVED** [review]
